### DISCUSSION TITLE
Exact match using PuMuPDF

### DISCUSSION THREAD

**Discussion by AnweshGangula** [pymupdf/PyMuPDF#1277] [Q&A]
I'm working on an app in which we need to verify if a text is an exact match (eg: "city" doesn't match "electricity"). 
I've looked at few existing questions( #678  & [Stack-Overflow](https://stackoverflow.com/questions/64536027/selecting-the-exact-match-using-pymupdf-page-searchfor)) but couldn't find an accurate solution.

I found a way that is the most accurate out of all the solutions available online (below). 

But I'm facing an issue when trying to verify text in multiple lines

Here is my code:

```python
def isExactMatch(page, term, clip):
    termLen = len(term)
    termBboxLen = max(abs(clip[0]-clip[2]),abs(clip[1]-clip[3]))
    termfontSize = termBboxLen/termLen
    f = termfontSize*2

    validate = page.get_text("blocks", clip = clip + (-f, -f, f, f), flags=0)[0][4]

    matches = len(re.findall(f'\\b{term}\\b', validate))>0

    return matches
```

This works perfectly fine, except in multi-line searches - I think its because the `clip` in `page.get_text` isn't spanning across multiple line

Is there any way to make it more accurate?

**Comment by JorjMcKie** [comment]
Within a text block you could check whether the exact sequence of the word(s) of your search string is equal to a sub-sequence of the block text split into words ... Just an idea.

Or extract the text as "words", sort them by block-number, line-number, vertical and horizontal coordinates. Then look wether your ``needle.split()`` matches a sub-sequence of the overall sorted word list ...

**Reply by AnweshGangula** [reply]
This may not exactly work for my use-case because if there are multiple instances of the needle in the page, I need to verify if each instance is an exact match.
This doesn't work for this because, it will be true if at-least one instance is found.

**Reply by JorjMcKie** [reply]
But you can also restrict the "words" output to clips.

**Reply by JorjMcKie** [reply]
or only consider those words that have the same block-number as your previously used block

**Reply by AnweshGangula** [reply]
I actually am using the clip right now to compare ( _you can check the `validate` variable in code above_), but for some reason, the clip isn't accurate when the text is multi-line. The output with clip is only the text before the line-break.

And the block-number is new information for me. is there any resource that explains what a block-number is? I would like to give that a try

**Reply by JorjMcKie** [reply]
1. where does your clip come from?
2. block number is contained as a positional value in each item of text output for "blocks" and "words" - see docu.

**Reply by AnweshGangula** [reply]
the clip is from `page.search_for(term)`

**Reply by JorjMcKie** [reply]
I thought so, but how do you compute it?
The response consists of multiple rectangles usually, which you must join to compute your clip. Every response may contain multiple rectangles for a single needle hit! For example if the word ``needle = "method"`` happens to be hyphenated into "meth-" and "od" across a line break, you will get two rectangles. Even more so this is true for needles made from multiple words.
So, when you do that:
How do you determine that only one occurrence of your needle is contained in the response ... and you do not build a clip wrapping all those occurrences?

**Reply by AnweshGangula** [reply]
oh, I didn't realize that I get 2 rectangles for hyphenated text across line breaks. Just checked the same.
But for needles made for multiple words, I'm getting a single rectangle (if its in same line)

This is the reason, it wasn't working as expected. I'll give the block-number approach a try and let you know.

**Reply by JorjMcKie** [reply]
yeah, if, if ...
What I mean is this: suppose a PDF page like this:
![grafik](https://user-images.githubusercontent.com/8290722/134375266-99371e73-786c-434f-8015-ac548e4a5fac.png)
And you search for ``needle = "for this first needle"``. Then this happens:
```
>>> rl=page.search_for(needle)
>>> len(rl)
3
>>> for r in rl:
	print(page.get_textbox(r))

	
for this
first needle
for this first needle
>>> 
```
My suggestions to try a solution based on words had this complication in mind.

**Reply by JorjMcKie** [reply]
It might be easier to check whether:
```python
>>> needle.split()
['for', 'this', 'first', 'needle']
>>> # is a sublist of this list:
>>> page.get_text().split()
['I', 'am', 'looking', 'for', 'this', 'first', 'needle,', 'then', 'for', 'the', 'second', 'needle,', 'and', 'so', 'on.', 'Once', 'I', 'am', 'finished', 'looking', 'for', 'this', 'first', 'needle,', 'I', 'need', 'to', 'make', 'sure', 'that', 'I', 'am', 'not', 'computing', 'my', 'clip', 'as', 'too', 'big', 'a', 'rectangle.']
```
And if so how many times ...

**Reply by JorjMcKie** [reply]
Also please do read the docu for details of what may be the response of `page.search_for()` in the several corner cases.
You may conclude that it is not always the appropriate choice.

**Reply by AnweshGangula** [reply]
Thank you @JorjMcKie , 
Actually, what i'm trying to do is - add highlights in the PDF based on a list of needles. And before adding the highlights, checking if its an exact match.

And I found another similar discussion where you mention we can add a list of quads to highlight text in multiple lines [#341 ](https://github.com/pymupdf/PyMuPDF/issues/341#issuecomment-519580903). This is what I'm looking for, but I would need to now identify which quads are from the same needle to pass them as a list.

Like in you're screenshot above, how do I combine the quads of "for this" and "first needle"

**Reply by AnweshGangula** [reply]
I think this would be a good feature to add to the `page.search_for()` method - to group the quads or rectangles based on the needle.

**Comment by JorjMcKie** [comment]
> Like in you're screenshot above, how do I combine the quads of "for this" and "first needle"

1. Combining **_quads_** is a difficult thing - explaining this later. Combining rectangles in contrast is easy: `union = rect1 | rect2`.
2. Rectangle `union` in this case is however equal to a rectangle covering the complete first two rows! Don't know if that helps you.

Combining arbitrary quads may not make any sense, because
* a quad may be an **_arbitrary_** tetragon, i.e. not be rectangular
* two **_rectangular_** quads may not be rotated by the same angle relative to the x-axis

In both above cases it would be hard to even **_uniquely_** define what the joining quad should be. You can however easily build the envelopping rectangle of a quad (`Quad.rect`) and join it with other rectangles as mentioned.

Also please read the docu for how to highlight text which potentially is spread across multiple lines and you know the top-left of the begin and bottom-right of the end.

> but I would need to now identify which quads are from the same needle

Where is the problem? If one of the rectangles ``.startswith`` the first word of needle you have found the start. Then check the following rectangles until you encounter one that ``.endswith`` the last word of needle.
